### ISSUE TITLE
[Inductor] torch.compile changes output shape of x.norm(dim=-1)

### ISSUE THREAD

**Issue by michael-960** [pytorch/pytorch#182181]
### 🐛 Describe the bug

`norm()` behaves differently before and after torch.compile

```python

class TestModule(torch.nn.Module):
    def forward(self, x):
        return x.norm(dim=-1)


m = TestModule()
n = torch.compile(m)

x = torch.rand(8, 3, 1)

print(m(x).shape)
print(n(x).shape) 
```

I get the following output:

```
torch.Size([8, 3])
torch.Size([8, 3, 1]) 
```

### Versions

Vulnerability Tsx async abort:           Mitigation; TSX disabled
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] pytorch-lightning==2.6.1
[pip3] torch==2.11.0
[pip3] torchmetrics==1.9.0
[pip3] torchvision==0.26.0
[pip3] triton==3.6.0
[conda] Could not collect

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by aorenste** [comment]
Looks like inductor - problem goes away using backend="eager"

**Comment by liqiangxl** [comment]
## Root Cause
- **File(s)**: `torch/_refs/linalg/__init__.py:168-187` — `vector_norm()` function
- **Root cause**: The `vector_norm` decomposition has a fast-path optimization for when all reduced dimensions have size 1. Instead of the full `pow → sum → pow` computation, it just takes `abs(x)` and reshapes. The reshape logic on line 186 uses:
  ```python
  new_shape = [s for d, s in enumerate(x.shape) if d not in dim]
  ```
  `enumerate(x.shape)` produces positive indices `(0, 8), (1, 3), (2, 1)`, but `dim` can contain negative values like `[-1]`. Since `2 != -1`, the check `d not in dim` is always True, and no dimension is removed from the output shape.
- **Call chain**: `torch.compile` → Inductor decomposition → `aten.linalg_vector_norm` → `torch/_refs/linalg/__init__.py:vector_norm()` → size-1 fast path → incorrect shape filtering
- **Why the guard on line 174 works**: `x.shape[d]` with `d=-1` uses Python's native negative indexing, so the condition `x.shape[-1] == 1` correctly evaluates to True and the fast path is entered. The bug is only in the shape construction afterward.

## Fix
- **Changed files**: `torch/_refs/linalg/__init__.py` (added 2 lines)
- **Approach**: Added `dim = utils.canonicalize_dims(x.ndim, dim)` after the `dim == []` check and before the size-1 fast-path guard. This converts negative indices to positive (e.g. `[-1]` → `(2,)` for a 3D tensor), so the `d not in dim` comparison works correctly.
- **Diff summary**:
  ```python
  if dim == []:
      dim = None
  elif dim is not None:
      dim = utils.canonicalize_dims(x.ndim, dim)  # NEW
  ```
